ipam/ipam.go

```diff
@@ -0,0 +1,223 @@
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostClusterName must not be empty", config)
+	}
+
+	s := &IPAM{
+		hostFramework: config.HostFramework,
+		logger:        config.Logger,
+		provider:      config.Provider,
+
+		commonDomain:    config.CommonDomain,
+		hostClusterName: config.HostClusterName,
+	}
+
+	return s, nil
+}
+
+func (c *IPAM) Test(ctx context.Context) error {
+
+	var (
+		clusterOne   = c.hostClusterName + "-cluster0"
+		clusterTwo   = c.hostClusterName + "-cluster1"
+		clusterThree = c.hostClusterName + "-cluster2"
+		clusterFour  = c.hostClusterName + "-cluster3"
+
+		// allocatedSubnets[clusterName]subnetCIDRStr
+		allocatedSubnets = make(map[string]string)
+		err              error
+	)
+
+	defer func() {
+		c.logger.LogCtx(ctx, "level", "debug", "message", "ensuring all guest clusters possibly created in test are deleted.")
+
+		for _, cn := range []string{clusterOne, clusterTwo, clusterThree, clusterFour} {
+			c.provider.DeleteCluster(cn)
+		}
+	}()
+
+	clusters := []string{clusterOne, clusterTwo, clusterThree}
+
+	c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating three guest clusters: %#v", clusters))
+
+	for _, cn := range clusters {
+		err = c.provider.CreateCluster(cn)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("create guest cluster %s", cn))
+	}
+
+	guestFrameworks := make(map[string]*framework.Guest)
+
+	c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("waiting for three guest clusters to be ready: %#v", clusters))
+
+	for _, cn := range clusters {
+		cfg := framework.GuestConfig{
+			Logger: c.logger,
+
+			ClusterID:    cn,
+			CommonDomain: c.commonDomain,
+		}
+		guestFramework, err := framework.NewGuest(cfg)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		guestFrameworks[cn] = guestFramework
+		err = guestFramework.Setup()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("guest cluster %s ready", cn))
+
+		awsConfig, err := c.hostFramework.AWSCluster(cn)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", "verify that there are no duplicate subnet allocations")
+		subnet := awsConfig.Status.Cluster.Network.CIDR
+		otherCluster, exists := allocatedSubnets[subnet]
+		if exists {
+			return microerror.Maskf(alreadyExistsError, "subnet %s already exists for %s", subnet, otherCluster)
+		}
+
+		// Verify that allocated subnets don't overlap.
+		for k, _ := range allocatedSubnets {
+			err := verifyNoOverlap(subnet, k)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+
+		allocatedSubnets[subnet] = cn
+	}
+
+	c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("terminating guest cluster %s and immediately creating new guest cluster %s", clusterTwo, clusterFour))
+
+	c.provider.DeleteCluster(clusterTwo)
```

Comment:
The test steps are still kind of wild and it is pretty hard to read and thus understand the test and how it works. So far we have a pretty clear layout where we have the `{ }` code blocks and opening and closing logs so all functionality is described and encapsulated. I miss this here and would appreciate to continue with this consistently. It makes it way more easy to read and understand the test and makes it less "wild" and "messy".

Reply:
Roger. I'll work on this.